graphqlapi/src/main/java/io/stargate/graphql/web/resources/schemafirst/NamespaceResource.java

```diff
@@ -0,0 +1,141 @@
+import org.slf4j.LoggerFactory;
+
+/** Serves the GraphQL schema deployed on each namespace. */
+@Singleton
+@Path(ResourcePaths.NAMESPACES)
+@Produces(MediaType.APPLICATION_JSON)
+@Authenticated
+public class NamespaceResource extends GraphqlResourceBase {
+
+  private static final Logger LOG = LoggerFactory.getLogger(NamespaceResource.class);
+  private static final Pattern NAMESPACE_PATTERN = Pattern.compile("\\w+");
+
+  @Inject private SchemaFirstCache graphqlCache;
+
+  @GET
+  @Path("/{namespace}")
+  public void get(
+      @PathParam("namespace") String namespace,
+      @QueryParam("query") String query,
+      @QueryParam("operationName") String operationName,
+      @QueryParam("variables") String variables,
+      @Context HttpServletRequest httpRequest,
+      @Suspended AsyncResponse asyncResponse) {
+
+    GraphQL graphql = getGraphql(namespace, httpRequest, asyncResponse);
+    if (graphql != null) {
+      get(query, operationName, variables, graphql, httpRequest, asyncResponse);
+    }
+  }
+
+  @POST
+  @Path("/{namespace}")
+  @Consumes(MediaType.APPLICATION_JSON)
+  public void postJson(
+      @PathParam("namespace") String namespace,
+      GraphqlJsonBody jsonBody,
+      @QueryParam("query") String queryFromUrl,
+      @Context HttpServletRequest httpRequest,
+      @Suspended AsyncResponse asyncResponse) {
+
+    GraphQL graphql = getGraphql(namespace, httpRequest, asyncResponse);
+    if (graphql != null) {
+      postJson(jsonBody, queryFromUrl, graphql, httpRequest, asyncResponse);
+    }
+  }
+
+  @POST
+  @Path("/{namespace}")
+  @Consumes(APPLICATION_GRAPHQL)
+  public void postGraphql(
+      @PathParam("namespace") String namespace,
+      String query,
+      @Context HttpServletRequest httpRequest,
+      @Suspended AsyncResponse asyncResponse) {
+
+    GraphQL graphql = getGraphql(namespace, httpRequest, asyncResponse);
+    if (graphql != null) {
+      postGraphql(query, graphql, httpRequest, asyncResponse);
+    }
+  }
+
+  private GraphQL getGraphql(
+      String namespace, HttpServletRequest request, AsyncResponse asyncResponse) {
+    if (!NAMESPACE_PATTERN.matcher(namespace).matches()) {
+      LOG.warn("Invalid namespace in URI, this could be an XSS attack: {}", namespace);
+      // Do not reflect back the value
+      replyWithGraphqlError(Response.Status.BAD_REQUEST, "Invalid namespace name", asyncResponse);
+      return null;
+    }
+
+    try {
+      GraphQL graphql =
+          graphqlCache.getGraphql(namespace, RequestToHeadersMapper.getAllHeaders(request));
+      if (graphql == null) {
+        replyWithGraphqlError(
+            Response.Status.NOT_FOUND,
+            String.format(
+                "Could not find a GraphQL schema for '%s', either the namespace does not exist, "
+                    + "or no schema was deployed to it yet.",
+                namespace),
+            asyncResponse);
+        return null;
+      } else if (!isAuthorized(request, namespace)) {
```

Comment:
Why not authorize before checking? Checking first could leak whether something does or doesn't exist.

Reply:
IIRC we had that discussion for the CQL-first API. If we don't want to reveal whether it's missing vs forbidden, then we would have to pick a single error code for both conditions.

Side note: the user is already authentified at this point.

Reply:
I see what you mean now, it's done differently in the CQL-first resource. This will get taken care of when I merge the two resources.